pq-crypto/s2n_kyber_512_evp.c

```diff
@@ -20,57 +20,62 @@
 
 #include "error/s2n_errno.h"
 #include "tls/s2n_kem.h"
+#include "utils/s2n_safety.h"
 #include "utils/s2n_safety_macros.h"
 
 #if defined(S2N_LIBCRYPTO_SUPPORTS_KYBER512)
-int s2n_kyber_512_evp_generate_keypair(uint8_t *public_key, uint8_t *private_key) {
-    EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, NULL);
+
+DEFINE_POINTER_CLEANUP_FUNC(EVP_PKEY *, EVP_PKEY_free);
+DEFINE_POINTER_CLEANUP_FUNC(EVP_PKEY_CTX *, EVP_PKEY_CTX_free);
+
+int s2n_kyber_512_evp_generate_keypair(uint8_t *public_key, uint8_t *secret_key) {
+    DEFER_CLEANUP(EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, NULL), EVP_PKEY_CTX_free_pointer);
     POSIX_GUARD_PTR(kyber_pkey_ctx);
-    POSIX_ENSURE(EVP_PKEY_keygen_init(kyber_pkey_ctx), S2N_FAILURE);
+    POSIX_ENSURE(EVP_PKEY_CTX_kem_set_params(kyber_pkey_ctx, NID_KYBER512_R3), S2N_FAILURE);
 
     EVP_PKEY *kyber_pkey = NULL;
```

Comment:
Do you need a `DEFER_CLEANUP` for this struct?

Reply:
no, [`EVP_PKEY_CTX_kem_set_params`](https://github.com/aws/aws-lc/blob/1d4332266a171fa41b5ae50faeb9c4df11187847/include/openssl/evp.h#L884) doesn't create a struct, it just sets the KEM function on the `kyber_pkey_ctx`, which is created using a `DEFER_CLEANUP` on L32.